src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala

```diff
@@ -81,94 +81,71 @@ trait TreeAndTypeAnalysis extends Debugging {
+      val tpApprox = analyzer.approximateAbstracts(tp)
+      val pre      = tpApprox.prefix
+      val sym      = tp.typeSymbol
+
+      def subclassesToSubtypes(syms: List[Symbol]): List[Type] = syms.flatMap { sym =>
+        // have to filter out children which cannot match: see ticket #3683 for an example
+        // compare to the fully known type `tp` (modulo abstract types),
+        // so that we can rule out stuff like:
+        //    sealed trait X[T]; class XInt extends X[Int]
+        // XInt not valid when enumerating X[String]
+        // however, must also approximate abstract types
+        val memberType  = nestedMemberType(sym, pre, tpApprox.typeSymbol.owner)
+        val subTp       = appliedType(memberType, WildcardType.fillList(sym.typeParams.length))
+        val subTpApprox = analyzer.approximateAbstracts(subTp)
+        if (subTpApprox <:< tpApprox) Some(checkableType(subTp)) else None
+      }
 
-            groupChildren(sym :: Nil, Nil)
-          } else {
-            val subclasses = debug.patmatResult(s"enum $sym sealed, subclasses")(
-              // symbols which are both sealed and abstract need not be covered themselves, because
-              // all of their children must be and they cannot otherwise be created.
-              sym.sealedDescendants.toList
-                sortBy (_.sealedSortName)
-                filterNot (x => (x.isSealed || x.isPrivate) && x.isAbstractClass && !isPrimitiveValueClass(x))
-            )
-
-            List(debug.patmatResult(s"enum sealed tp=$tp, tpApprox=$tpApprox as") {
-              // valid subtypes are turned into checkable types, as we are entering the realm of the dynamic
-              filterChildren(subclasses)
-            })
-          }
-        case sym if sym.isCase =>
-          List(List(tp))
+      def filterAndSortChildren(children: Set[Symbol]) = {
+        // symbols which are both sealed and abstract need not be covered themselves,
+        // because all of their children must be and they cannot otherwise be created.
+        children.toList.filter(x => !(x.isSealed || x.isPrivate) || !x.isAbstractClass || isPrimitiveValueClass(x))
+          .sortBy(_.sealedSortName)
+      }
 
-        case sym =>
-          debug.patmat("enum unsealed "+ ((tp, sym, sym.isSealed, isPrimitiveValueClass(sym))))
-          Nil
+      @tailrec def groupChildren(wl: List[Symbol], acc: List[List[Symbol]]): List[List[Symbol]] = wl match {
+        case Nil      => acc
+        case hd :: tl =>
+          val children = filterAndSortChildren(hd.sealedChildren)
+          // put each trait in a new group since traits could belong to the same group as a derived class
+          val (traits, nonTraits) = children.partition(_.isTrait)
+          groupChildren(tl ::: children, acc ::: traits.map(List(_)).appended(nonTraits))
+      }
+
+      val subclasses = debug.patmatResult(s"enum $sym sealed, subclasses") {
+        if (grouped) groupChildren(List(sym), Nil)
```

Comment:
This is the only remaining use of grouped. Assuming both branches return the same elements but grouped differently, is it worth it to eliminate grouped altogether and always make it true? If so, there are two call sites that call it with false, of which one would have to be adapted not to use headOption

Reply:
Yeah, I wanted to too.  But I didn't want to get into trouble with the allocation police.

Reply:
The allocation police will eventually come by and change those `:::`'s to mutable builders anyway